src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs

```diff
@@ -0,0 +1,94 @@
+        context.RegisterCompilationStartAction(OnCompilationStarted);
+    }
+
+    private static void OnCompilationStarted(CompilationStartAnalysisContext context)
+    {
+        var analyzer = new AnalyzerImplementation(context.Compilation);
+        if (analyzer.IsValid)
+        {
+            context.RegisterSymbolAction(analyzer.AnalyzeOperationInvocation, SymbolKind.NamedType);
+        }
+    }
+
+    private sealed class AnalyzerImplementation
+    {
+        private readonly INamedTypeSymbol? _domainEventType;
+        private readonly INamedTypeSymbol? _domainEventAttributeType;
+        private readonly ImmutableHashSet<INamedTypeSymbol> _domainEventTypes;
+
+        public AnalyzerImplementation(Compilation compilation)
+        {
+            this._domainEventType = compilation.GetBestTypeByMetadataName(KnownSymbolNames.DomainEventInterface, KnownSymbolNames.WorkleapDomainEventPropagationAbstractionsAssembly)!;
+            this._domainEventAttributeType = compilation.GetBestTypeByMetadataName(KnownSymbolNames.DomainEventAttribute, KnownSymbolNames.WorkleapDomainEventPropagationAbstractionsAssembly)!;
+
+            var domainEventPropagationSymbols = ImmutableHashSet.CreateBuilder<INamedTypeSymbol>(SymbolEqualityComparer.Default);
+            domainEventPropagationSymbols.AddIfNotNull(this._domainEventType);
+            domainEventPropagationSymbols.AddIfNotNull(this._domainEventAttributeType);
+
+            this._domainEventTypes = domainEventPropagationSymbols.ToImmutable();
+        }
+
+        public bool IsValid => this._domainEventTypes.Count == 2 &&
+                               this._domainEventType is not null &&
+                               this._domainEventAttributeType is not null;
+
+        public void AnalyzeOperationInvocation(SymbolAnalysisContext context)
+        {
+            if (context.Symbol is INamedTypeSymbol { TypeKind: TypeKind.Class or TypeKind.Struct, IsAbstract: false } type)
+            {
+                if (this.IsDomainEventClass(type))
+                {
+                    var hasDomainEventAttribute = type.GetAttributes()
+                        .Any(x => x.AttributeClass != null && SymbolEqualityComparer.Default.Equals(x.AttributeClass, this._domainEventAttributeType));
+
+                    if (!hasDomainEventAttribute)
+                    {
+                        context.ReportDiagnostic(UseDomainEventAttribute, type);
+                    }
+                }
+            }
+        }
+
+        private bool IsDomainEventClass(ITypeSymbol type)
+        {
+            if (this.ImplementsBaseDomainEventInterface(type))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool ImplementsBaseDomainEventInterface(ITypeSymbol type)
+        {
+            return type.AllInterfaces.Any(x => SymbolEqualityComparer.Default.Equals(x, this._domainEventType));
+        }
```

Comment:
It doesn't seem necessary to have two methods, we could only keep the last one (change the name if you want to)